### ISSUE TITLE
Provide a way to run tests against an existing database

### ISSUE THREAD

**Issue by bittner** [mixxorz/behave-django#11]
Sometimes in integration testing it may make sense to run BDD tests against an existing database, such as a copy of a production database. -- This is fundamentally different to unit tests where we always want a predefined state for testing against a predefined result.

At the moment this is not possible with Django and behave-django, because the `django_test_runner` creates and destroys, and the underlying `LiveServerTestCase` modifies and flushes the database used for testing.

The solution may provide a command line switch, e.g. `--no-test-setup`, that makes clear that the current default settings are used, identical to running `python manage.py runserver` directly.

**Comment by mixxorz** [comment]
Will exporting the production database as a fixture and loading it before every test satisfy this requirement?

If we do add a `--no-test-setup` flag, that would mean created database objects won't reset every after test. I don't know how that would be useful...

**Comment by bittner** [comment]
> created database objects won't reset every after test. I don't know how that would be useful...
1. There's quite a large part of test cases in BDD that acts kind of "read only", testing data being displayed correctly.
2. Changes may not need to be undone, because we're on a CI, a testing or a development system.
3. The database is restored easily by sync'ing it from the reference system.

## tl;dr

Imagine web agencies that don't do automated testing yet. For each and every project they of course have a handover-phase with manual acceptance testing. How do they get that job done?
1. There are test procedures, written down probably in Word documents, that describe the steps to execute and the expected result.
2. The developers prepare a testing or staging environment where the (manual) tests are carried out.
3. There may be no automatic initialization of the test environment, and certainly there is no database being reset after each test.
4. Restoring a clean test environment is the developers' or DevOps' job. They ideally have a script that does it all in a single step (e.g. `fab sync_database sync_media`).

Allowing them to use their usual test setup, while helping them to turn their test procedures into Gherkin test specs, can help them to get a big step closer to test automation.

**Comment by mixxorz** [comment]
What's preventing them from testing normally? Come up with a starting test "state" and put that into a fixture which will get loaded every time you test. That's actually the point of fixtures.

From the Django docs.

> It’s sometimes useful to pre-populate your database with hard-coded data when you’re first setting up an app. There’s a couple of ways you can have Django automatically create this data: you can provide initial data via fixtures, or you can provide initial data as SQL.

If there really is a use case for it though, I don't mind putting it in. Make a PR.

**Comment by bittner** [comment]
The Django docs talk about unit tests. The complete mind set in testing is about unit tests. And I agree that with unit tests this is the one and only way, really.

## Tearing down the walls

As described above it's much about tearing down the walls between the no-time-for-writing-tests and the i-write-tests-for-my-living people. Believe it or not, there's still a vast amount of software being developed without tests. And this is because business people have a hard time seeing the advantages of test-driven (BDD) development, and programmers have a hard time explaining why "let's do it without tests, that's faster" is a really bad choice.

## Start testing. Your way!

Why not give them both a hand? -- How do Django developers do their front-end facing work? Isn't it usually writing or modifying templates (and the babies around it), and try out the results on a local runserver? Okay. Now, let's tell them to write the test first (with semantics in mind!), and _then_ implement the template. Great! And soon they'll stop testing every single change manually, and start running `python manage.py behave` instead. Cool!

Then, gradually, we'll tell them more about best practices. Which will get us closer to the strict unit test mind set, with test data specifically prepared (which requires more effort, but may give more reliable results).

## More choice. More fun. More change possible.

It's about paving the road towards investing in quality software. It's about allowing people to move towards developing tested software very much their way. Letting people do it "their way" avoids resistance towards change. Less resistance means less frustration, and less frustration means "let's do more of these cool things!"

So, I'll prepare a PR. Unfortunately, this is the hard part. Because Django's unit tests are not designed for that kind of freedom. :weary:

**Comment by mixxorz** [comment]
I understand that you want to make it easier for people to get started with testing their code base. You want them to do proper TDD and BDD. That's all well and good.

But your last comment still didn't explain why you want your tests to _not_ run in isolation.

This is the workflow I have in my head
1. Export the production database into a fixture.
2. Add code to load the fixture before all tests.
3. Write your tests as if you _are_ running on the production database. (Because they have the same data anyway)

The only difference I could think of is that with this method, tests don't affect other tests, which is a very good thing. It's going to be super difficult to write tests for a system where you don't have a predefined starting state, not to mention being a bad practice.

**Comment by bittner** [comment]
I agree that the workflow you describe _is_ good practice. There are several flaws, though, for test beginners.
- Importing a complete database from fixtures (a full dumpdata) is not a trivial thing. You'll have [some trouble to go though](http://stackoverflow.com/questions/10350559/django-fixtures-from-dumpdata-failing-when-testing) before you get it right.
- Your local changes are not reflected in the (reference) database. If you have new models you can add a migration, granted. If you configure things (in your local dev database) you can add additional fixtures for that, correct.

What I'm saying is, again, it makes exploratory development less easy. It forces developers immediately to invest more for doing it "the right way". And they won't be willing to do it. Not immediately.

I firmly believe in best practices. I really do. But they need to enter when good practices are already standard, and have replaced the bad ones. We're talking about making it _easier_. That's it.
